src/socket_base.cpp

```diff
@@ -1289,6 +1295,7 @@ int zmq::socket_base_t::close ()
     //  process.
     send_reap (this);
 
+    EXIT_MUTEX ();
```

Comment:
in general the EXIT_MUTEX should be before returning from the method.

read-after-free can only happen in non-thread-safe socket and then it doesn't matter as the mutex is not being used. however I understand we need to solve it, I think adding the following to the socket_base destructor can solve the problem:

```C
sync.lock ();
sync.unlock ();
```

Another solution is to actually remove the mutex from the close method, leaving it unsafe by design (which kind of unsafe).

Also we can just move the EXIT MUTEX before the send_reap, I don't see much harm in that.